### PR TITLE
apply to all accounts in the AWS Organization

### PR DESCRIPTION
**Pull request by afeld**
https://trello.com/c/U1hu7CNH/339-as-everyone-i-want-a-newly-brokered-aws-account-to-have-the-networking-set-up

### CODE REVIEW

**Comment by afeld** [comment]
Ok, got this up to proof-of-concept stage. My thinking is that we want to ensure all tenant accounts have their networking (and really, any required base configuration we decide on) kept up-to-date at all times. The code here loops through all accounts in the Organization and `apply`s the Terraform configuration to each independently. This could be done on a recurring basis from Lambda, CircleCI, or wherever.

The separate states are kept in Terraform [workspaces](https://www.terraform.io/docs/state/workspaces.html) corresponding to each account. It treats each account independently; if one account fails to update for some reason, it logs that, but then continues on to the next one.

Here's the output, with the account/user IDs replaced with dummy values:

```
$ python apply.py
Running for account 11111.
Workspace '11111' exists.
+ terraform workspace select 11111
Switched to workspace "11111".
+ terraform apply -input=false -auto-approve
data.aws_caller_identity.current: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

account_id = 11111
caller_arn = arn:aws:sts::11111:assumed-role/OrganizationAccountAccessRole/22222
caller_user = SOMEUSER:22222

Successfully set up networking for account 11111.
---------
Running for account 33333.
Workspace '33333' exists.
+ terraform workspace select 33333
Switched to workspace "33333".
+ terraform apply -input=false -auto-approve
data.aws_caller_identity.current: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

account_id = 33333
caller_arn = arn:aws:sts::33333:assumed-role/OrganizationAccountAccessRole/44444
caller_user = OTHERUSER:44444

Successfully set up networking for account 33333.
---------
Networking set up for all accounts successfully.
```

Rather than [refactoring the networking code to be account-agnostic](https://trello.com/c/aoCmkiPN/336-refactor-tenant-networking-test-code-to-use-multiple-accounts-now-that-aws-account-broker-is-working), I created a `test/` module that distills the idea down. This is only temporary for the proof-of-concept.

Let me know if any of that is unclear, and what you think!

**Comment by tim20watts** [comment]
I would like to see some governance around this process before
the proverbial summit button is ever hit. For something like this we need
ISSO sign off on each FISMA system that the change will happen in. And I
believe we need to set precedence of time, day and a change windows of when
these standard changes will take place. and also work on putting together
processes for CCBs.

Thank you in advance,
Timothy E. Watts - IDIHC
Innovation Project Management Office - Innovation Specialist
Office (504) 249-6438 | Cell (985) 237-0782 | timothy.watts@gsa.gov |
www.gsa.gov

**Comment by afeld** [comment]
> For something like this we need ISSO sign off on each FISMA system that the change will happen in

You mean the tenant systems? Since they are inheriting controls from our networking, I'm not sure that's the case.

> we need to set precedence of time, day and a change windows of when these standard changes will take place

My thinking is that this script would be automatically run on a schedule specified in code, so we get that for "free". 

> putting together processes for CCBs

This is a longer conversation, but curious to hear the outcomes/considerations you'd be looking for from a Change Control Board, rather than simply "we need one, because that's the way things have always been done". Trying to stay open-minded, despite the fact that I see Change Control Boards (as I hear of them existing today) as an antipattern.

File: apply.py [hunk 1]
```diff
@@ -27,9 +28,14 @@ def get_tenant_accounts():
             yield account
 
 def set_up_networking(role_arn):
+    my_env = os.environ.copy()
+    my_env['AWS_DEFAULT_REGION'] = 'us-east-1'
```
Comment (Vermyndax) [inline apply.py]: Do we really want to force us-east-1 on them?

**Comment by Vermyndax** [comment]
I like where this is going so far. The terraform pieces need some modifications to be broken up properly across accounts. As written, it only works within a single account with different regions. We'll need to do some work against that before this will really be a workable approach. Do we want to split it up into terraform runs, called by the python script individually, or do we want to have terraform executed once against all four accounts?

File: apply.py (hunk 1 above)
Comment (afeld) [reply apply.py]: Well, someone would need to pick a region(s) for the VPC etc. to be created in...thoughts about how else it could be done?

**Comment by afeld** [comment]
> The terraform pieces need some modifications to be broken up properly across accounts

https://trello.com/c/aoCmkiPN/336-refactor-tenant-networking-test-code-to-use-multiple-accounts-now-that-aws-account-broker-is-working

> Do we want to split it up into terraform runs, called by the python script individually, or do we want to have terraform executed once against all four accounts?

As you know, I'm strongly opposed to hard-coding the number of environments, so I would propose the script deal with accounts independently. In other words, the former 😄

**Comment by afeld** [comment]
Coming back to the pull request itself: @Vermyndax sounds like you're ok with the general approach, yeah? Should we merge and iterate from there, or are there other things you'd like to see first?

**Comment by Vermyndax** [comment]
@afeld I do like this approach. I think this would be a good Lambda function, which would work well since it's in Python. Maybe run the lambda function from the master account. We would need to govern the IAM roles to make sure they're not jerked around, but that's not impossible. How do we want to store tfvars for the various environments and maintain/govern the subnets and IP spaces? There's still some things to decide there, I think.

**Comment by Vermyndax** [comment]
@afeld Currently this branch has no terraform to apply against it, right? It's just the python script, correct?

**Comment by Vermyndax** [comment]
@afeld I think my only concern is that I'm not sure how the VPN connectivity will play with this (since I strongly suspect part of that will be manual on the transit VPC side) and I'm a little concerned about the VPC peering being difficult to use. We can discuss tomorrow. It could be that I just lack imagination on this right now.

**Comment by afeld** [comment]
> I think this would be a good Lambda function, which would work well since it's in Python.

Yeah, picked Python because I knew it would leave that option open.

> How do we want to store tfvars for the various environments and maintain/govern the subnets and IP spaces?

Dunno yet.

> I think my only concern is that I'm not sure how the VPN connectivity will play with this (since I strongly suspect part of that will be manual on the transit VPC side) and I'm a little concerned about the VPC peering being difficult to use.

I'm not sure either. I say we take that as a follow-up.

---

Good to merge?

**Comment by briankfitzwater** [comment]
FYI: [Go lang is supported on AWS Lambda](https://aws.amazon.com/blogs/compute/announcing-go-support-for-aws-lambda/)

**Pull request merged**